### ISSUE TITLE
`pyright-langserver>=1.1.335` LSP does not annotate parameters for completion

### ISSUE THREAD

**Issue by lmmx** [microsoft/pyright#6734]
I used pyright LSP for the first time the other night, then soon after put it on another machine. I used my base conda Python environment, which had other things installed in it, so there were constraints that meant the 2 machines got different versions.

I was surprised to find that the 2nd machine I installed it on got more features! The main feature was that on the 2nd machine (v1.1.333) I could type

```py
from pydantic import BaseModel, ConfigDict

class A(BaseModel):
    model_config = ConfigDict(
```

and then get a completion e.g. `validate_default=...)`

# Spec differences

- Both machines: Linux Mint, conda, Python 3.11, vim 8
- Machine 1: v1.1.340, no type completions for imported TypedDict subclass keys as kwargs
- Machine 2: v1.1.333, with type completions for imported TypedDict subclass keys as kwargs

I inspected the environment requirements with `pip freeze`, then made fresh conda installations that successfully reproduced the bug, detailed below.

## Reproduction

- I installed `vim-lsp` and `vim-lsp-settings` which together automate the setup of LSPs in vim.

```vim
call plug#begin('~/.vim/plugged')

Plug 'skywind3000/asyncrun.vim'
Plug 'prabirshrestha/asyncomplete.vim'
Plug 'prabirshrestha/vim-lsp'
Plug 'mattn/vim-lsp-settings'
Plug 'prabirshrestha/asyncomplete-lsp.vim'

" Initialize plugin system
call plug#end()

" Tab completion for asyncomplete.vim
" https://github.com/prabirshrestha/asyncomplete.vim
inoremap <expr> <Tab>   pumvisible() ? "\<C-n>" : "\<Tab>"
inoremap <expr> <S-Tab> pumvisible() ? "\<C-p>" : "\<S-Tab>"
inoremap <expr> <cr>    pumvisible() ? asyncomplete#close_popup() : "\<cr>"

" Tone down the visual distraction of the 'sign gutter' from LSP linters
let g:lsp_diagnostics_enabled = 0
let g:lsp_document_highlight_enabled = 0

let g:lsp_settings_filetype_python = ['ruff-lsp', 'pyright-langserver']

" Configure the Pyright and Python LSP servers
let g:lsp_settings = {
\   'ruff-lsp': {
\     'disabled': v:false,
\     'workspace_config': {
\       'python': {}
\     }
\   },
\   'pyright-langserver': {
\     'disabled': v:false,
\     'workspace_config': {
\       'python': {
\         'analysis': {
\           'useLibraryCodeForTypes': v:true,
\         }
\       }
\     }
\   }
\}
```

Note: it's fine to have the setting for ruff-lsp alongside, if it's not installed it'll just report as "exited"

- I used `node`/`npm` via conda and via nvm but it made no difference (both will reproduce the bug, I think conda is v19 and nvm v20)

I made many attempts to narrow down the exact version at which this behaviour appeared. Each time I created a fresh miniconda env (`~/miniconda3`) which is self-contained and would mean I didn't need to change the `~/.bashrc` setting which made the Python binary (and thus `site-packages`) there available in the shell.

- [miniconda installer direct link](https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh)
  - [via](https://docs.conda.io/projects/miniconda/en/latest/)

### Note on reproducing

- the LSP server takes a little time to set up on first installation, you can run `:LspStatus` which will either show "starting", "running", or "exited". if it's still "starting", close the file and retry in a few seconds.

### Last known good version: 1.1.334

Installing the following requirements, `reqs_INIT_minimal_even_later.txt`:

```
pyright==1.1.334
pydantic==2.5.2
```

We have parameter type completion and documentation 'bubble' (box, popup, frame?)

```py
from pydantic import BaseModel, ConfigDict



class A(BaseModel):
    model_config = ConfigDict(
    a: int = 1               ╔══════════════════════════════════════════════════════
    b: int = 2               ║(`__map: ConfigDict`, /, *, title: str | None = ...,  
                             ║str_to_lower: bool = ..., str_to_upper: bool = ..., s 
                             ║tr_strip_whitespace: bool = ..., str_min_length: int  
a = A()                      ║= ..., str_max_length: int | None = ..., extra: Extra 
                             ║Values | None = ..., frozen: bool = ..., populate_by_ 
a.model_dump()               ║name: bool = ..., use_enum_values: bool = ..., valida 
~                            ║te_assignment: bool = ..., arbitrary_types_allowed: b 
~                            ║ool = ..., from_attributes: bool = ..., loc_by_alias: 
```

### First known bad version: 1.1.335

Installing the following requirements, `reqs_INIT_minimal_too_late.txt`:

```
pyright==1.1.335
pydantic==2.5.2
```

You don't get any parameter types, just `*args: Any`, `**kwargs: Any`:

```py
from pydantic import BaseModel, ConfigDict
                              ╔═══════════════════════════════════════════════╗
                              ║(`*args: Any`, **kwargs: Any) -> ConfigDict    ║
                              ║A TypedDict for configuring Pydantic behaviour.║
class A(BaseModel):           ╚═══════════════════════════════════════════════╝
    model_config = ConfigDict(
```

This matches the behaviour I saw on v1.1.340 and appears to be the earliest reproducible version, so I would expect to find the cause in the changelog for this version.

## Logging

I added logging to `.vimrc`:

```vim
" for vim-lsp log
let g:lsp_log_verbose = 1
let g:lsp_log_file = expand('~/tmp/lsp/vim-lsp.log')

" for asyncomplete.vim log
let g:asyncomplete_log_file = expand('~/tmp/lsp/asyncomplete.log')
```

Attached are the 2 log files each in the normal format and a linted JSON format (in the case of `vim-lsp.log.json` dates removed), logging the few seconds in which I opened the file `foo.py` (roughly equal to the demo code block above), went to line 5, opened a bracket after ConfigDict, saw the lack of parameter annotations, and then exited the file.

- [vim-lsp.log](https://github.com/microsoft/pyright/files/13673106/vim-lsp.log) 351 lines
- [asyncomplete.log](https://github.com/microsoft/pyright/files/13673107/asyncomplete.log) 113 lines

as JSON:
- [vim-lsp.log.json](https://github.com/microsoft/pyright/files/13673104/vim-lsp.log.json) 13407 lines
- [asyncomplete.log.json](https://github.com/microsoft/pyright/files/13673105/asyncomplete.log.json) 5843 lines

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

This was due to a long-standing but in the language server's "signature help" support. It wasn't properly handling constructor calls where the `__init__` method was overloaded. An unrelated change introduced in 1.1.335 exposed this flaw for TypedDict constructor calls.

**Comment by lmmx** [comment]
Ah you already found what caused it, great to hear!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.341, which I just published.

**Comment by lmmx** [comment]
Confirmed fixed :trophy:
